regression_em/model.py

```diff
@@ -1,18 +1,258 @@
+        :param intercept: Intercept parameter of LR.
+        :param feat_vec: Feature vector.
+        :return: Sequence of probabilities.
+        """
+        # 1 / exp(- (feature @ coef + intercept))
+        return 1 / (1 + np.exp(- feat_vec @ coef - intercept))
+
+    def calc_logits(self, probs: np.ndarray) -> np.ndarray:
+        """Return logits calculated from probability array.
+
+        :param probs: Probability array to be converted to logists.
+        :param epsilon: Tiny value for clipping.
+        :return: Logist arrray.
+        """
+        # Perform clipping to avoid log(0) and zero division.
+        clipped = probs
+        clipped[clipped == 0] = self.epsilon
+        clipped[clipped == 1] = 1 - self.epsilon
+
+        return np.log(clipped / (1 - clipped))
+
+    @staticmethod
+    def calc_responsibility(target_prob: float, ref_prob: float, is_positive: bool) -> float:
+        """Return responsibility to be used in EM algorithm.
+        For detail, see eq.1 of [https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf].
+
+        :param target_prob: Probability calculated from the parameters to be updated.
+        :param ref_prob: Reference probability.
+        :param is_positive: If True, the sample has a positive label.
+        :return: Responsibility.
+        """
+
+        if is_positive or target_prob == ref_prob == 1:
+            return 1.0
+        else:
+            return (target_prob * (1 - ref_prob)) / (1 - target_prob * ref_prob)
+
+    def update_responsibilities(self, target_params: RegParam, target_feat: np.ndarray,
+                                ref_params: RegParam, ref_feat: np.ndarray,
+                                labels: Sequence[bool]) -> np.ndarray:
+        """Return responsibilities based on M-step parameters.
+
+        .. Note::
+            target_params and ref_params must be (weight vector, intercept).
+
+        :param target_params: Regression params corresponding to the latent factor to be updated.
+        :param target_feat: Feature matrix corresponding to the latent factor to be updated.
+        :param ref_params: Regression params corresponding to the refered latent factor.
+        :param ref_feat: Feature matrix corresponding to the refered latent factor.
+        :param labels: Sequence of boolean indicating each sample is positivei or negative.
+        :return: List of responsibilities.
+        """
+        # The format of params must be (coef vector, intercept).
+        # Calculating ref prob. with updated ref params for updating target param
+
+        target_probs = self.calc_probs(target_params[0], target_params[1], target_feat)
+        ref_probs = self.calc_probs(ref_params[0], ref_params[1], ref_feat)
+
+        return np.vectorize(self.calc_responsibility)(target_probs, ref_probs, labels)
+
+    def update_params(self, feat_mat: np.ndarray, responsibilities: np.ndarray, sample_weights) -> Tuple[RegParam, float]:
```

Comment:
RegParam is a tuple of np.array and float.

Reply:
Fixed it.